### PR TITLE
[EHealth] Add examples of Chinese Biomedical Language Understanding tasks

### PR DESCRIPTION
**Pull request by LemonNoel**
### PR types

New features

### PR changes

Models

### Description

- Add build-in parameters of Chinese pre-trained language model [EHealth](https://arxiv.org/abs/2110.07244)
- Add datasets in [CBLUE](https://github.com/CBLUEbenchmark/CBLUE) for sequence classification tasks to ``paddlenlp.dataset``
- Add sequence classification example 
- Add Exponential Moving Average implementation to ``paddlenlp.op.optimizer``

**TODO**
- Add examples of token classification and SPO prediction
- Add references of datasets

### CODE REVIEW

File: examples/biomedical/cblue/sequence_classification/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# 使用医疗领域预训练模型Fine-tune完成中文医疗文本分类任务
+
+近年来，预训练语言模型（Pre-trained Language Model，PLM）逐渐成为自然语言处理（Natural Language Processing，NLP）的主流方法。这类模型可以利用大规模的未标注语料进行训练，得到的模型在下游NLP任务上效果明显提升，在通用领域和特定领域均有广泛应用。在医疗领域，早期的做法是在预先训练好的通用语言模型上进行Fine-tune。后来的研究发现直接使用医疗相关语料学习到的预训练语言模型在医疗文本任务上的效果更好，采用的模型结构也从早期的BERT演变为更新的RoBERTa、ALBERT和ELECTRA。
+
+本示例展示了中文医疗预训练模型eHealth（[Building Chinese Biomedical Language Models via Multi-Level Text Discrimination](https://arxiv.org/abs/2110.07244)）如何Fine-tune完成中文医疗文本分类任务。
+
+## 模型介绍
+
+本项目针对中文医疗文本分类任务，开源了中文医疗预训练模型eHealth（简写`chinese-ehealth`）。eHealth（[Building Chinese Biomedical Language Models via Multi-Level Text Discrimination](https://arxiv.org/abs/2110.07244)）使用了医患对话、科普文章、病历档案、临床病理学教材等脱敏中文语料进行预训练，通过预训练任务设计来学习词级别和句级别的文本信息。该模型的整体结构与ELECTRA相似，包括生成器和判别器两部分。 而Fine-tune过程只用到了判别器模块，由12层Transformer网络组成。
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/README.md]: eHealth 不是ELECTRA Health 或者 ERNIE Health吗？

Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/README.md]: CBLUE相关信息是不是可以写上去了？

File: examples/biomedical/cblue/sequence_classification/README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+## 快速开始
+
+### 代码结构说明
+
+以下是本项目主要代码结构及说明：
+
+```text
+sequence_classification/
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/README.md]: classification 是不是就行了？ 

或者不要拆分开一个文件

File: examples/biomedical/cblue/sequence_classification/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+### 模型训练
+
+我们以中文医疗文本数据集CBLUE中的文本分类数据集为示例数据集，包括：
+
+* CHIP-CDN：给定病历档案，预测其中包含的规范化诊断实体。本项目使用了检索后重新构建的二分类数据集，给定病历档案和规范化诊断实体，预测前者是否包含后者（简写`CHIP-CDN-2C`）。
+* CHIP-CTC：给定医疗文本描述，按照中国临床筛选标准进行分类。
+* CHIP-STS：给定两个涉及5种不同疾病的句子，预测二者语义是否相似。
+* KUAKE-QIC：给定医疗问句，对患者咨询目的进行分类。
+* KUAKE-QTR：给定医疗问句和文章标题，预测二者内容是否一致。
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/README.md]: 医疗问句和文章标题 内容一致？

File: examples/biomedical/cblue/sequence_classification/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+* KUAKE-QQR：给定两个医疗问句，预测二者描述内容是否一致。
+
+可以运行下边的命令，在训练集上进行训练，并在开发集上进行验证。
+```shell
+$ unset CUDA_VISIBLE_DEVICES
+$ python -m paddle.distributed.launch --gpus "0" train.py --dataset CHIP-CDN-2C --batch_size 256 --max_seq_length 32 --weight_decay 0.01 --warmup_proportion 0.1
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/README.md]: 这里有个疑问，所有的数据集的超参都是一样的吗？

File: examples/biomedical/cblue/sequence_classification/train.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    'CHIP-STS': partial(
+        MultiLabelsMetric, name='macro'),
+    'CHIP-CDN-2C': partial(
+        MultiLabelsMetric, name='micro')
+}
+
+# yapf: disable
+parser = argparse.ArgumentParser()
+parser.add_argument('--dataset', choices=['KUAKE-QIC', 'KUAKE-QQR', 'KUAKE-QTR', 'CHIP-STS', 'CHIP-CTC', 'CHIP-CDN-2C'],
+                                 default='KUAKE-QIC', type=str, help='Dataset for token classfication tasks.')
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/train.py]: sequence classification

File: examples/biomedical/cblue/sequence_classification/train.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+parser.add_argument('--seed', default=1000, type=int, help='Random seed for initialization.')
+parser.add_argument('--device', choices=['cpu', 'gpu', 'xpu', 'npu'], default='gpu', help='Select which device to train model, default to gpu.')
+parser.add_argument('--epochs', default=3, type=int, help='Total number of training epochs to perform.')
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/train.py]: Total number of training epochs.

File: examples/biomedical/cblue/sequence_classification/train.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+parser.add_argument('--batch_size', default=32, type=int, help='Batch size per GPU/CPU for training.')
+parser.add_argument('--learning_rate', default=6e-5, type=float, help='Learning rate for fine-tuning sequence classification task.')
+parser.add_argument('--weight_decay', default=0.01, type=float, help="Weight decay if we apply some.")
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/train.py]: 这里说一下是Optimizer的weight decay?

File: examples/biomedical/cblue/sequence_classification/train.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+parser.add_argument('--warmup_proportion', default=0.1, type=float, help='Linear warmup proportion over the training process.')
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/train.py]: 这里也说一下是学习率的Linear warmup proportion

File: examples/biomedical/cblue/sequence_classification/train.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        paddle.distributed.init_parallel_env()
+
+    set_seed(args.seed)
+
+    train_ds, dev_ds, test_ds = load_dataset(
+        'cblue', args.dataset, splits=['train', 'dev', 'test'])
+
+    model = ppnlp.transformers.ElectraForSequenceClassification.from_pretrained(
+        'chinese-ehealth', num_classes=len(train_ds.label_list))
+    tokenizer = ppnlp.transformers.ElectraTokenizer.from_pretrained(
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/train.py]: 这里的待商榷，是否使用EHealth相关的class，来突出EHealth

File: examples/biomedical/cblue/sequence_classification/train.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        'chinese-ehealth')
+
+    trans_func = partial(
+        convert_example,
+        tokenizer=tokenizer,
+        max_seq_length=args.max_seq_length)
+    batchify_fn = lambda samples, fn=Tuple(
+        Pad(axis=0, pad_val=tokenizer.pad_token_id),  # input
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/train.py]: Pad这块设置一下dtype="int64", 防止windows Numpy默认编码int32的问题

File: examples/biomedical/cblue/sequence_classification/train.py [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+        dev_ds,
+        mode='dev',
+        batch_size=args.batch_size,
+        batchify_fn=batchify_fn,
+        trans_fn=trans_func)
+
+    if args.init_from_ckpt and os.path.isfile(args.init_from_ckpt):
+        state_dict = paddle.load(args.init_from_ckpt)
+        model.set_dict(state_dict)
+    model = paddle.DataParallel(model)
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/train.py]: 这里如果不是多卡，是不是就不用DataParallel

File: examples/biomedical/cblue/sequence_classification/train.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    criterion = paddle.nn.loss.CrossEntropyLoss()
+    if METRIC_CLASSES[args.dataset] is Accuracy:
+        metric = METRIC_CLASSES[args.dataset]()
+        metric_name = 'accuracy'
+    else:
+        metric = METRIC_CLASSES[args.dataset](
+            num_labels=len(train_ds.label_list))
+        metric_name = metric._name + ' f1'
+    if args.use_amp:
+        scaler = paddle.amp.GradScaler(init_loss_scaling=args.scale_loss)
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/train.py]: 这里也可以尝试一下 Pure Fp16

File: examples/biomedical/cblue/sequence_classification/utils.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unicodedata
+import numpy as np
+
+
+def _normalize(text):
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/utils.py]: 这块能不能内置到PaddleNLP Library里面了

File: examples/biomedical/cblue/sequence_classification/utils.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+                output.append(c)
+            output.append(' ')
+        elif cp == 0xF979:  # https://www.zhihu.com/question/20697984
+            output.append('凉')
+        else:
+            output.append(char)
+    return "".join(output)
+
+
+def _tokenize_special_chars(text):
```
Comment (wawltor) [inline examples/biomedical/cblue/sequence_classification/utils.py]: 这个函数应该Library应该是有的

File: paddlenlp/transformers/model_utils.py [hunk 15]
```diff
@@ -194,6 +195,7 @@ def from_pretrained(cls, pretrained_model_name_or_path, *args, **kwargs):
 
         # From built-in pretrained models
         if pretrained_model_name_or_path in pretrained_models:
+            cls.pretrained_model_name = pretrained_model_name_or_path
```
Comment (wawltor) [inline paddlenlp/transformers/model_utils.py]: 这里主要是干啥了

File: paddlenlp/datasets/cblue.py [hunk 16]
```diff
@@ -0,0 +1,339 @@
+import collections
```
Comment (wawltor) [inline paddlenlp/datasets/cblue.py:15]: 这里加一下Copyright

File: paddlenlp/ops/optimizer/ema.py [hunk 17]
```diff
@@ -0,0 +1,34 @@
+class ExponentialMovingAverage(object):
```
Comment (wawltor) [inline paddlenlp/ops/optimizer/ema.py:16]: 加一下Copyright

File: examples/biomedical/cblue/sequence_classification/README.md (hunk 1 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/README.md]: 这里是参照了[相应论文](https://arxiv.org/abs/2110.07244)中的名称

File: examples/biomedical/cblue/sequence_classification/README.md (hunk 4 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/README.md]: 超参是有不同的，之后把最终结果和相应的超参整理成表格放上去

File: paddlenlp/transformers/model_utils.py (hunk 15 above)
Comment (LemonNoel) [reply paddlenlp/transformers/model_utils.py]: 这个是为了直接在ElectraPretrainedModel里按照pretrained_model_name（例如'chinese_ehealth'或者'chinese_electra_base'）设置layernorm的epsilon和activation这两个参数，但是这么实现逻辑上和理解上感觉都有点问题，是不是直接改成函数的入参传进去比较好呢

File: paddlenlp/ops/optimizer/ema.py (hunk 17 above)
Comment (LemonNoel) [reply paddlenlp/ops/optimizer/ema.py:16]: Done.

File: paddlenlp/datasets/cblue.py (hunk 16 above)
Comment (LemonNoel) [reply paddlenlp/datasets/cblue.py:15]: Done.

File: examples/biomedical/cblue/sequence_classification/README.md (hunk 1 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/README.md]: OK

File: examples/biomedical/cblue/sequence_classification/README.md (hunk 2 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/README.md]: Done. 已将文本分类任务代码移到cblue目录下。

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 5 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/train.py]: Done.

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 6 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/train.py]: Done.

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 7 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/train.py]: Done.

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 8 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/train.py]: Done.

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 9 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/train.py]: 已线下讨论。这里与论文中的eHealth保持一致。

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 10 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/train.py]: Done.

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 11 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/train.py]: 已增加判断条件。

File: paddlenlp/transformers/model_utils.py (hunk 15 above)
Comment (LemonNoel) [reply paddlenlp/transformers/model_utils.py]: 已删除，改为传参。

File: examples/biomedical/cblue/sequence_classification/README.md (hunk 1 above)
Comment (ZeyuChen) [reply examples/biomedical/cblue/sequence_classification/README.md]: 是eHealth

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 9 above)
Comment (ZeyuChen) [reply examples/biomedical/cblue/sequence_classification/train.py]: 模型名是ehealth-chinese比较好还是还是chinese-ehealth？其他模型的格式是否有个标准？

File: examples/biomedical/cblue/sequence_classification/train.py [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+    rank = paddle.distributed.get_rank()
+    if paddle.distributed.get_world_size() > 1:
+        paddle.distributed.init_parallel_env()
+
+    set_seed(args.seed)
+
+    train_ds, dev_ds, test_ds = load_dataset(
+        'cblue', args.dataset, splits=['train', 'dev', 'test'])
+
+    model = ppnlp.transformers.ElectraForSequenceClassification.from_pretrained(
```
Comment (ZeyuChen) [inline examples/biomedical/cblue/sequence_classification/train.py]: 后续不要再用ppnlp.transformer这个缩写吧，直接在最顶层改为 
`from paddlenlp.transformer import ElectraForSequenceClassification`
否则实在是太长了

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 12 above)
Comment (ZeyuChen) [reply examples/biomedical/cblue/sequence_classification/train.py]: 这些配置后续需要Trainer化来降低成本

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 9 above)
Comment (ZeyuChen) [reply examples/biomedical/cblue/sequence_classification/train.py]: 我看了下模型名，还是建议以ehealth开头比较好，重点突出模型名。

File: examples/biomedical/cblue/sequence_classification/utils.py (hunk 13 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/utils.py]: Done.

File: examples/biomedical/cblue/sequence_classification/utils.py (hunk 14 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/utils.py]: 找了一下这个Library里并没有，已内置进去。

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 18 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/train.py]: 好的，已修改。

File: examples/biomedical/cblue/sequence_classification/train.py (hunk 9 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/train.py]: 了解。模型名已修改为ehealth-chinese.

File: examples/biomedical/cblue/sequence_classification/README.md (hunk 3 above)
Comment (LemonNoel) [reply examples/biomedical/cblue/sequence_classification/README.md]: 这里想表达的意思是描述的内容一致。更新后的README中去掉了这些自己写的数据集描述，增加了CBLUE的链接导向原始数据集的详细介绍。
